### DISCUSSION TITLE
Choosing a module system

### DISCUSSION THREAD

**Discussion by r0nsha** [r0nsha/chili-lang#21] [Ideas]
The current module system for Chili is based on the file system.

Imports are expressions returning the module they import, using a structured path, for example: `import foo/bar`.
Imports are not bound to a name automatically, this is done by the user, and with some powerful patterns, you can do some cool things like this: `let bar @ { baz, * } = import foo/bar`.
Also, visibility is centered around bindings, meaning that the bindings themselves are either public/private. Modules can always be imported, meaning that they are always public. This is a slight downside in my opinion.

In general, the thing that I'm missing the most in this system, is that while the import syntax itself is structured, the module tree isn't. You can freely import any module that you'd like.

I've been looking at other programming languages, and here's a summary of what I've seen.

**Rust:**

Although quite complicated, I really like their module system. It's based on the file system, which is great, but modules themselves are declared via a `mod` declaration. If you want a module to be public, you type `pub mod ...`. Although a bit boilerplate-y, this is quite intuitive in my opinion. This also brings the benefit that modules themselves have a visibility.

**Odin:**

Odin went with a simpler route, packages. Each file has to declare the package that it's in at the top. This is has the pro (and con) of being very flat. All packages are visible to the outside world, and visibility is determined at the binding level, which is similar to Chili.

**Zig:**

Zig uses imports as expressions, which actually return a `comptime` struct of the module's content. That's kinda neat!
Also, visibility in Zig is pretty straight forward as well, and it is based at the binding level.

**Story time:** Import expressions were inspired by Zig. Having binding pattern's be orthogonal between structs and modules was very appealing, so we went with that. Actually, the first module system was based on Rust, and imports looked something like this: `use foo.bar.*`.

Having imports as expressions is a nice thing that I'm not willing to let go very easily, unless it becomes so unergonomic that we have to ditch it. Having a single syntax for destructing and wildcarding is very powerful in my opinion.

This is a discussion to choose which direction we should go with this? Do you just go the "virtual filesystem" route of Rust, and stick with expression based imports?

Do you go with a full rewrite, and make `import` a separate statement, with its own syntax and everything?

Or do we go even further, and introduce packages to Chili?

Or do you have another idea or suggestion?

**Comment by ruarq** [comment]
I'd pick the first option
> Do you go with a full rewrite, and make import a separate statement, with its own syntax and everything?

I like the way Rust solves the module problem too. Packages feel kind of Java-ish, and I don't like having to declare for each file which package it's in.

I think leveraging the tree properties of the file system is the way to go.

**Reply by r0nsha** [reply]
Thanks for replying!

And what about the import syntax?
Would you prefer writing the first or the second?

1. `let * = import foo`
2. `import foo.*`

Or even:
3. `use foo.*`

**Reply by ruarq** [reply]
I'd prefer `import foo.*`.

`use foo.*` is pretty short, but coming from C++ I'd say `use` / `using` / `...` is something I'd associate with a type alias or a `using namespace` declaration.

`let * = import foo` looks like a variable declaration, and while it can be neat to import a module into a "variable", I think it's weird writing `let *` if you want to import the module just like that, without a name.

I think python has a good solution for importing modules. I don't quite agree with syntax, but overall it's good.
I especially don't like the additional `from` and `as` keywords, but there's many different ways you could solve that.

One additional problem I see with `import foo.*` is the redundant repetition of `foo.` if you want to import multiple specific symbols: `import foo.baz, foo.baz`, but maybe you already have a solution for that :)

**Reply by r0nsha** [reply]
For the issue you raised with variable declaration, that's actually what it does. This could be confusing though for people coming from other languages.

As for repetitiveness, both the current syntax and the alternate `import foo` syntax have solutions for this:

In this current syntax you can do:
```
let { bar, baz } = import foo;

let { look, this: { is: { deep } } } = import hey;
```

Btw, the same exact syntax applies to structs in this version.

In the new syntax it would probably something like:
```
import foo.{ bar, baz };

import hey.{ look, this.is.deep };
```

**Reply by magicmouse** [reply]
Modula2 which i consider the high water mark for module systems, had an interesting property in their IMPORT syntax, in that all imports were required to be at the beginning of each module, so that a aystem that builds the dependency graph (the makefile generator you might call it), can run with minimal scanning of source code.

A very clever syntactical restriction that breeds efficiency.

M2 had the concept of opaque record exports, whereby you could name something but not be able to access the content of a record (i.e. get into the fields).  

JS bungled its import system, and although you can import a scalar from another module, you cannot modify it.

**Reply by r0nsha** [reply]
@magicmouse That's interesting. 

I don't think that restricting declaration order is not something that I'd be willing to do. 

Also, JS's modules are very underwhelming.

The main dilemmas here are:
1. How do we express module privacy?
2. Is `import` an expression, or a statement? If it stays as an expression, do we add some syntax sugar to make it more familiar to users coming from other languages?

**Reply by magicmouse** [reply]
Modules allow a person to build up a larger system out of smaller pieces, and keep logically related chunks of code in tidy piles, like having a math library.  If very module was independent and pure, this would make it easy to create a module system. But modules are squishy sometimes, and module A can call B which can call C which can call A again.  Do you permit cycles between modules? 

Zig is making module imports a compile time thing, so that you can build a MAKE tool in Zig that will peruse the total graph of modules in a project and then compile them. 

There is the issue of introspection of modules. Some languages have elaborate introspection schemes. In Beads i create a big tree data structure which is augmented by each module, and thus you finally once initialization is completed of all modules, have a full database of the functions, variables, definitions, etc. that are in each module.  This is useful for building IDE tools, so that a program can peek into the symbols of another.

There is further the issue of what order are the module initialization bodies executed in?  How is it specified? Does the system do topological sort, or is random?  Many languages are a bit murky as to the fine points on these critical issues. Once a project gets large enough it can have 100-200 modules, and then cross-module issues really show up.
